### PR TITLE
Fix Data race in semi-join

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description

As described in https://github.com/vitessio/vitess/issues/17410 and https://github.com/vitessio/vitess/issues/17411, there is a data race in the stream execute code of semi-join.

This PR adds a test to uncover the problem and fixes it.

## Related Issue(s)

- Fixes https://github.com/vitessio/vitess/issues/17410
- Fixes https://github.com/vitessio/vitess/issues/17411

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

**Review by systay — APPROVED** [review]

File: go/vt/vtgate/engine/semi_join.go [hunk 1]
```diff
@@ -62,24 +63,26 @@ func (jn *SemiJoin) TryExecute(ctx context.Context, vcursor VCursor, bindVars ma
...
 	err := vcursor.StreamExecutePrimitive(ctx, jn.Left, bindVars, wantfields, func(lresult *sqltypes.Result) error {
+		joinVars := make(map[string]*querypb.BindVariable)
 		result := &sqltypes.Result{Fields: lresult.Fields}
 		for _, lrow := range lresult.Rows {
 			for k, col := range jn.Vars {
 				joinVars[k] = sqltypes.ValueBindVariable(lrow[col])
 			}
-			rowAdded := false
+			var rowAdded atomic.Bool
 			err := vcursor.StreamExecutePrimitive(ctx, jn.Right, combineVars(bindVars, joinVars), false, func(rresult *sqltypes.Result) error {
```
Comment (harshit-gangal) [inline go/vt/vtgate/engine/semi_join.go:74]: There is another issue I mentioned: during a transaction, we might end up opening two connections, leaving one of them in limbo.
Do you have another PR that fixes it?

Comment (GuptaManan100) [reply go/vt/vtgate/engine/semi_join.go:74]: I will open a separate issue for it, with separate tests and PR for it, after this one gets merged.

File: go/vt/vtgate/engine/semi_join.go [hunk 2]
```diff
@@ -62,24 +63,26 @@ func (jn *SemiJoin) TryExecute(ctx context.Context, vcursor VCursor, bindVars ma
...
-					rowAdded = true
+				if len(rresult.Rows) > 0 {
+					rowAdded.Store(true)
 				}
 				return nil
 			})
 			if err != nil {
 				return err
 			}
+			if rowAdded.Load() {
```
Comment (dbussink) [inline go/vt/vtgate/engine/semi_join.go:83]: Why is this one atomic? If this needs an atomic since it otherwise races, I imagine that the `append` on the next line also races and needs to be guarded with a lock then? 

If this doesn't race, it doesn't need to be atomic?

Comment (GuptaManan100) [reply go/vt/vtgate/engine/semi_join.go:83]: Yes i know, I made the changes such that i could avoid using this atomic thing, but in the callback right above we are writing to it. We are only setting it to true but golang still complains in a `-race` test saying there is a concurrent write. So I had to make it atomic to avoid that. That being said, for strictly correctness purposes, we didn't need to make it atomic, but I couldn't figure out how to make the test work without making it atomic.

File: go/vt/vtgate/engine/semi_join_test.go [hunk 3]
```diff
@@ -159,3 +159,60 @@ func TestSemiJoinStreamExecute(t *testing.T) {
...
+	jn := &SemiJoin{
+		Left:  leftPrim,
+		Right: rightPrim,
+		Vars: map[string]int{
+			"bv": 1,
+		},
+	}
+	err := jn.TryStreamExecute(context.Background(), &noopVCursor{}, map[string]*querypb.BindVariable{}, true, func(result *sqltypes.Result) error {
+		return nil
+	})
```
Comment (harshit-gangal) [inline go/vt/vtgate/engine/semi_join_test.go:213]: should the results be validated?

File: go/vt/vtgate/engine/semi_join.go (hunk 2 above)
Comment (harshit-gangal) [reply go/vt/vtgate/engine/semi_join.go:83]: both are racy as the internal `StreamExecute` can do parallel execute to multiple shards

Comment (dbussink) [reply go/vt/vtgate/engine/semi_join.go:83]: @harshit-gangal Right, so that means we need to guard the `append` too right? Or otherwise we lose data there potentially?

File: go/vt/vtgate/engine/semi_join_test.go (hunk 3 above)
Comment (GuptaManan100) [reply go/vt/vtgate/engine/semi_join_test.go:213]: Okay.

File: go/vt/vtgate/engine/semi_join.go (hunk 2 above)
Comment (GuptaManan100) [reply go/vt/vtgate/engine/semi_join.go:83]: No we don't. The Result field is local to each outer StreamExecute call. Its initialized in the callback.

Comment (harshit-gangal) [reply go/vt/vtgate/engine/semi_join.go:83]: results is local to outer callback but append happens inside the second callback, which itself can execute in parallel causes race in append.

Comment (harshit-gangal) [reply go/vt/vtgate/engine/semi_join.go:83]: If you really want to test that, you would need a test that does scatter call for the inside stream execute (right side of semi-join)

File: go/vt/vtgate/engine/semi_join.go (hunk 1 above)
Comment (harshit-gangal) [reply go/vt/vtgate/engine/semi_join.go:74]: would be good to create an issue to not miss it

File: go/vt/vtgate/engine/semi_join.go (hunk 2 above)
Comment (GuptaManan100) [reply go/vt/vtgate/engine/semi_join.go:83]: Append is not inside the second callback. That is the entire point.

Comment (GuptaManan100) [reply go/vt/vtgate/engine/semi_join.go:83]: The test I added does a scatter for the inside semi-join and the outside semi-join too.

File: go/test/endtoend/vtgate/queries/misc/misc_test.go [hunk 4]
```diff
@@ -519,3 +519,24 @@ func TestTimeZones(t *testing.T) {
 		})
 	}
 }
+
+// TestSemiJoin tests that the semi join works as intended.
+func TestSemiJoin(t *testing.T) {
+	utils.SkipIfBinaryIsBelowVersion(t, 22, "vtgate")
```
Comment (harshit-gangal) [inline go/test/endtoend/vtgate/queries/misc/misc_test.go:525]: do we need this now?

File: go/vt/vtgate/engine/semi_join.go (hunk 2 above)
Comment (harshit-gangal) [reply go/vt/vtgate/engine/semi_join.go:83]: This looks good.

**Review by harshit-gangal — APPROVED** [review]

**Review by dbussink — APPROVED** [review]

File: go/test/endtoend/vtgate/queries/misc/misc_test.go (hunk 4 above)
Comment (GuptaManan100) [reply go/test/endtoend/vtgate/queries/misc/misc_test.go:525]: Yes, the test is still flaky for older releases, although it fails very infrequently.

Comment (harshit-gangal) [reply go/test/endtoend/vtgate/queries/misc/misc_test.go:525]: we do not run v22 test on v21 VTGate in upgrade/downgrade test.
To really skip this test, this should be in `release-21` branch than on `main`

We have already planned the backport. So that will not be an issue.

Comment (GuptaManan100) [reply go/test/endtoend/vtgate/queries/misc/misc_test.go]: I removed the line.
